crates/bevy_ecs/src/system/commands/entity_command.rs

```diff
@@ -0,0 +1,329 @@
+/// #   fn as_str(&self) -> &str { &self.0 }
+/// # }
+///
+/// #[derive(Resource, Default)]
+/// struct Counter(i64);
+///
+/// /// A `Command` which names an entity based on a global counter.
+/// fn count_name(entity: Entity, world: &mut World) {
+///     // Get the current value of the counter, and increment it for next time.
+///     let mut counter = world.resource_mut::<Counter>();
+///     let i = counter.0;
+///     counter.0 += 1;
+///
+///     // Name the entity after the value of the counter.
+///     world.entity_mut(entity).insert(Name::new(format!("Entity #{i}")));
+/// }
+///
+/// // App creation boilerplate omitted...
+/// # let mut world = World::new();
+/// # world.init_resource::<Counter>();
+/// #
+/// # let mut setup_schedule = Schedule::default();
+/// # setup_schedule.add_systems(setup);
+/// # let mut assert_schedule = Schedule::default();
+/// # assert_schedule.add_systems(assert_names);
+/// #
+/// # setup_schedule.run(&mut world);
+/// # assert_schedule.run(&mut world);
+///
+/// fn setup(mut commands: Commands) {
+///     commands.spawn_empty().queue(count_name);
+///     commands.spawn_empty().queue(count_name);
+/// }
+///
+/// fn assert_names(named: Query<&Name>) {
+///     // We use a HashSet because we do not care about the order.
+///     let names: HashSet<_> = named.iter().map(Name::as_str).collect();
+///     assert_eq!(names, HashSet::from_iter(["Entity #0", "Entity #1"]));
+/// }
+/// ```
+///
+/// # Note on Generic
+///
+/// The `Marker` generic is necessary to allow multiple blanket implementations
+/// of `EntityCommand` for closures, like so:
+/// ```ignore (This would conflict with the real implementations)
+/// impl EntityCommand for FnOnce(Entity, &mut World)
+/// impl EntityCommand<World> for FnOnce(EntityWorldMut)
+/// impl EntityCommand<Result> for FnOnce(Entity, &mut World) -> Result
+/// impl EntityCommand<(World, Result)> for FnOnce(EntityWorldMut) -> Result
+/// ```
+/// Without the generic, Rust would consider the implementations to be conflicting.
+///
+/// The type used for `Marker` has no connection to anything else in the implementation.
+pub trait EntityCommand<Marker = ()>: Send + 'static {
+    /// Executes this command for the given [`Entity`] and
+    /// returns a [`Result`] for error handling.
+    fn apply(self, entity: Entity, world: &mut World) -> Result;
+
+    /// Returns a [`Command`] which executes this [`EntityCommand`] for the given [`Entity`].
+    ///
+    /// This method is called when adding an [`EntityCommand`] to a command queue via [`Commands`](crate::system::Commands).
+    /// You can override the provided implementation if you can return a `Command` with a smaller memory
+    /// footprint than `(Entity, Self)`.
+    /// In most cases the provided implementation is sufficient.
+    #[must_use = "commands do nothing unless applied to a `World`"]
+    fn with_entity(self, entity: Entity) -> impl Command<(Result, CommandError)>
```

Comment:
Same here as in my other comment - this strikes me as the sort of method which should probably return a wrapper struct (i.e. `WithEntity<Self>`).